sos/plugins/nvme.py

```diff
@@ -0,0 +1,119 @@
+## (at your option) any later version.
+
+## This program is distributed in the hope that it will be useful,
+## but WITHOUT ANY WARRANTY; without even the implied warranty of
+## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+## GNU General Public License for more details.
+
+## You should have received a copy of the GNU General Public License
+## along with this program; if not, write to the Free Software
+## Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+
+from sos.plugins import Plugin, RedHatPlugin, DebianPlugin, UbuntuPlugin
+import os, subprocess
+
+class NvmePlugin(Plugin, RedHatPlugin, DebianPlugin, UbuntuPlugin):
+    """Collect config and system information about NVMe devices"""
+
+    plugin_name = "nvme"
+    devices = [] 
+
+    def copy_spec(self):
+        """ Loop through all NVMe devices in /sys/block/<nvme-device>/ 
+	and get the files inside queue/, device/, integrity/ and mq/ """
+
+        i = 0
+        while os.path.isdir('/sys/block/nvme%dn1' % i):
+            self.devices.append("nvme%dn1" % i)
+
+            queue_files = subprocess.check_output("ls -1 /sys/block/nvme%dn1/queue/*" % i, shell=True).rstrip()
+            # in python 3.3 or above the prefix b' is not ignored, so bytes.decode() is used to get rid of it
+            queue_files = bytes.decode(queue_files)
+            queue_files = queue_files.splitlines()
+            for q in queue_files:
+                self.add_copy_spec(q)
+
+            device_files = subprocess.check_output("find /sys/block/nvme%dn1/device/ -maxdepth 1 -type f" % i, shell=True).rstrip()
+            device_files = bytes.decode(device_files)
+            device_files = device_files.splitlines()
+            for d in device_files:
+                self.add_copy_spec(d)
+
+            general_files = subprocess.check_output("find /sys/block/nvme%dn1/ -maxdepth 1 -type f" % i, shell=True).rstrip()
+            general_files = bytes.decode(general_files)
+            general_files = general_files.splitlines()
+            for g in general_files:
+                self.add_copy_spec(g)
+
+            integrity_files = subprocess.check_output("ls -1 /sys/block/nvme%dn1/integrity/*" % i, shell=True).rstrip()
+            integrity_files = bytes.decode(integrity_files)
+            integrity_files = integrity_files.splitlines()
+            for ing in integrity_files:
+                self.add_copy_spec(ing)
+
+            mq_files = subprocess.check_output("find /sys/block/nvme%dn1/mq/ -type f" % i, shell=True).rstrip()
+            mq_files = bytes.decode(mq_files)
+            mq_files = mq_files.splitlines()
+            for m in mq_files:
+                self.add_copy_spec(m)	
+
+            i += 1
+
+    def setup(self):
+        self.copy_spec()
+
+        for dev in self.devices:
+            # get block size
+            self.add_cmd_output("sh -c \"lsblk | grep %s | awk '{ print $4 }'\"" % dev, suggest_filename="block-size.%s" % dev)
```

Comment:
Nack. Either do this properly in Python or not at all - this is not a justification for using a shell pipeline.